File: Difficulty: Hard/Count Spanning Trees in a Graph/count-spanning-trees-in-a-graph.java
class Solution {
    public int countSpanTree(int n, int[][] edges) {
        if (n == 1) return 1;
        double[][] lap = new double[n][n];
        for (int[] e : edges) {
            int u = e[0], v = e[1];
            lap[u][u] += 1;
            lap[v][v] += 1;
            lap[u][v] -= 1;
            lap[v][u] -= 1;
        }
        int m = n - 1;
        double[][] mat = new double[m][m];
        for (int i = 1; i < n; i++) {
            for (int j = 1; j < n; j++) {
                mat[i - 1][j - 1] = lap[i][j];
            }
        }
        return (int) Math.round(determinant(mat));
    }
    private double determinant(double[][] a) {
        int n = a.length;
        double det = 1;
        for (int i = 0; i < n; i++) {
            int pivot = i;
            for (int j = i; j < n; j++) {
                if (Math.abs(a[j][i]) > Math.abs(a[pivot][i])) {
                    pivot = j;
                }
            }
            if (Math.abs(a[pivot][i]) < 1e-9) return 0;
            double[] temp = a[i];
            a[i] = a[pivot];
            a[pivot] = temp;
            if (i != pivot) det *= -1;
            det *= a[i][i];
            for (int j = i + 1; j < n; j++) {
                double factor = a[j][i] / a[i][i];
                for (int k = i; k < n; k++) {
                    a[j][k] -= factor * a[i][k];
                }
            }
        }
        return det;
    }
}

File: Difficulty: Hard/Count Spanning Trees in a Graph/count-spanning-trees-in-a-graph.cpp
class Solution {
private:
    double determinant(vector<vector<double>>& mat, int n) {
        double det = 1.0;
        const double EPS = 1e-9;

        for(int i = 0; i < n; ++i) {
            int pivot = i;
            for(int j = i + 1; j < n; ++j) {
                if(abs(mat[j][i]) > abs(mat[pivot][i])) {
                    pivot = j;
                }
            }

            if(abs(mat[pivot][i]) < EPS) return 0.0;

            if(i != pivot) {
                swap(mat[i], mat[pivot]);
                det = -det;
            }

            det *= mat[i][i];

            for(int j = i + 1; j < n; ++j) {
                double factor = mat[j][i] / mat[i][i];
                for(int k = i; k < n; ++k) {
                    mat[j][k] -= factor * mat[i][k];
                }
            }
        }
        return det;
    }

public:
    int countSpanTree(int n, vector<vector<int>>& edges) {
        if(n <= 1) return 1;

        vector<vector<double>> laplacian(n, vector<double>(n, 0.0));
        for(const auto& edge : edges) {
            int u = edge[0];
            int v = edge[1];
            laplacian[u][u] += 1.0;
            laplacian[v][v] += 1.0;
            laplacian[u][v] -= 1.0;
            laplacian[v][u] -= 1.0;
        }

        vector<vector<double>> submatrix(n - 1, vector<double>(n - 1, 0.0));
        for(int i = 1; i < n; ++i) {
            for(int j = 1; j < n; ++j) {
                submatrix[i - 1][j - 1] = laplacian[i][j];
            }
        }

        double totalTrees = determinant(submatrix, n - 1);
        return static_cast<int>(round(totalTrees));
    }
};